src/Roslyn.Diagnostics.Analyzers/Core/SymbolIsBannedAnalyzer.cs

```diff
@@ -0,0 +1,314 @@
+                    where StringComparer.Ordinal.Equals(Path.GetFileName(additionalFile.Path), BannedSymbolsFileName)
+                    let sourceText = additionalFile.GetText(compilationContext.CancellationToken)
+                    where sourceText != null
+                    from line in sourceText.Lines
+                    let text = line.ToString()
+                    where !string.IsNullOrWhiteSpace(text)
+                    select new ApiLine(text, line.Span, sourceText, additionalFile.Path);
+
+                var apiLines = query.ToList();
+
+                if (apiLines.Count == 0)
+                {
+                    return ImmutableHashSet<(ISymbol, string)>.Empty;
+                }
+
+                var lineById = new Dictionary<string, ApiLine>(StringComparer.Ordinal);
+                var errors = new List<Diagnostic>();
+                var builder = ImmutableHashSet.CreateBuilder<(ISymbol symbol, string message)>();
+
+                foreach (var line in apiLines)
+                {
+                    if (lineById.TryGetValue(line.DeclarationId, out var existingLine))
+                    {
+                        errors.Add(Diagnostic.Create(DuplicateBannedSymbolRule, line.Location, new[] { existingLine.Location }, line.DeclarationId));
+                        continue;
+                    }
+
+                    lineById.Add(line.DeclarationId, line);
+
+                    var symbols = DocumentationCommentId.GetSymbolsForDeclarationId(line.DeclarationId, compilationContext.Compilation);
+                    if (!symbols.IsDefaultOrEmpty)
+                    {
+                        foreach (var symbol in symbols)
+                        {
+                            builder.Add((symbol, line.Message));
+                        }
+                    }
+                }
+
+                if (errors.Count != 0)
+                {
+                    compilationContext.RegisterCompilationEndAction(
+                        endContext =>
+                        {
+                            foreach (var error in errors)
+                            {
+                                endContext.ReportDiagnostic(error);
+                            }
+                        });
+                }
+
+                return builder.ToImmutable();
+            }
+
+            void ReportDiagnostic(Diagnostic diagnostic)
+            {
+                diagnostics = diagnostics ?? new List<Diagnostic>(4);
+                diagnostics.Add(diagnostic);
+            }
+
+            void VerifyAttributes(ImmutableArray<AttributeData> attributes)
+            {
+                foreach (var attribute in attributes)
+                {
+                    if (bannedAttributes.TryGetValue(attribute.AttributeClass, out var message))
+                    {
+                        var node = attribute.ApplicationSyntaxReference.GetSyntax();
```

Comment:
Can this throw or return null for attributes from metadata?

Reply:
It does look like `ApplicationSyntaxReference` can return `null`, such as via `PEAttributeData`.

I think it's safe to just continue when we see `null`. It seems the value will always be non-`null` when we're inspecting source, which is the always the case when we want to report a diagnostic as far as I can tell.

I'm looking forward to C# 8 for this kind of thing :)

Fixed in 04b49422ab24ffc8f4195b7ff6b1b6b4fbf94d83.